gateway/handle_http_auth.go

```diff
@@ -119,6 +119,44 @@ func (gw *Handle) sourceIDAuth(delegate http.HandlerFunc) http.HandlerFunc {
 	}
 }
 
+// authenticateDestinationIdForSource middleware to authenticate destinationId in the X-Rudder-Destination-Id header.
+// If the destinationId is invalid, the request is rejected.
+// destinationID authentication should be performed only after source is authenticated and source is present in context
+// Following validations are performed
+//  1. Destination should be present for source config
+//  2. Destination should be enabled for the source
+func (gw *Handle) authenticateDestinationIdForSource(delegate http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		reqType := r.Context().Value(gwtypes.CtxParamCallType).(string)
+		arctx := r.Context().Value(gwtypes.CtxParamAuthRequestContext).(*gwtypes.AuthRequestContext)
+		var errorMessage string
+		defer func() {
+			gw.handleHttpError(w, r, errorMessage)
+			gw.handleFailureStats(errorMessage, reqType, arctx)
+		}()
+		destinationID := r.Header.Get("X-Rudder-Destination-Id")
+		if destinationID == "" {
+			if !gw.config.GetBool("Gateway.isDestinationIdMandatoryInRequestHeader", false) {
+				delegate.ServeHTTP(w, r)
+				return
+			}
+			errorMessage = response.NoDestinationIdInHeader
+			return
+		}
+		isValidDestination, isDestinationEnabled := gw.validateDestinationID(destinationID, arctx)
+		if !isValidDestination {
+			errorMessage = response.InvalidDestinationID
+			return
+		}
+		if !isDestinationEnabled {
+			errorMessage = response.DestinationDisabled
+			return
+		}
```

Comment:
We can probably simplify this. We probably don't need validateDestinationID.

```suggestion
		destination, found := lo.Find(arctx.Source.Destinations, func(destination backendconfig.DestinationT) bool {
			return destination.ID == destinationID
		})
		if !found {
			errorMessage = response.InvalidDestinationID
			return
		}
		if !destination.Enabled {
			errorMessage = response.DestinationDisabled
			return
		}
```

Reply:
IMO its better to separate out core logic in small functions, if in future we want to modify any logic we will only need to modify small functions.